File: 31.그래프와 순회/24480.알고리즘 수업 - 깊이 우선 탐색 2/Main.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Stack;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N, M, R;
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        R = Integer.parseInt(st.nextToken());

        List<List<Integer>> graph = new ArrayList<>();
        for (int i = 0; i <= N + 1; i++) {
            graph.add(new ArrayList<>());
        }

        for (int i = 0; i < M; i++) {
            int u, v;
            st = new StringTokenizer(br.readLine());
            u = Integer.parseInt(st.nextToken());
            v = Integer.parseInt(st.nextToken());
            graph.get(u).add(v);
            graph.get(v).add(u);
        }

        int[] isVisit = new int[N + 1];
        Stack<Integer> stack = new Stack<>();
        stack.add(R);

        int cnt = 1;
        while (!stack.isEmpty()) {
            int cur = stack.pop();
            if (isVisit[cur] != 0)
                continue;

            isVisit[cur] = cnt++;
            Collections.sort(graph.get(cur));
            for (int next : graph.get(cur)) {
                if (isVisit[next] != 0)
                    continue;

                stack.add(next);
            }
        }

        StringBuilder sb = new StringBuilder();
        for (int i = 1; i <= N; i++) {
            sb.append(isVisit[i]).append("\n");
        }
        System.out.println(sb.toString());
        br.close();

    }
}

File: 31.그래프와 순회/24480.알고리즘 수업 - 깊이 우선 탐색 2/main.py
import sys

input = sys.stdin.readline

if __name__ == "__main__":
    N, M, R = map(int, input().split())

    # 그래프 정보
    graph = [list() for _ in range(N + 1)]

    # 간선 연결 정보
    for _ in range(M):
        u, v = map(int, input().split())
        graph[u].append(v)
        graph[v].append(u)

    # DFS ---------
    stack = list()
    isVisit = [0 for _ in range(N + 1)]

    # 시작점 설정
    stack.append(R)

    # DFS 시작
    cnt = 1
    while stack:
        cur = stack.pop()
        if isVisit[cur] != 0:
            continue
        isVisit[cur] = cnt
        cnt += 1

        for next in sorted(graph[cur]):
            if isVisit[next] != 0:
                continue
            stack.append(next)

    # 출력
    print("\n".join(map(str, isVisit[1:])))
